certora/AllocatorVault.spec

```diff
@@ -0,0 +1,228 @@
+    assert vatInkVaultAfter == vatInkVaultBefore, "draw did not keep vat.urns(ilk,vault).ink unchanged";
+    assert vatArtVaultAfter == vatArtVaultBefore + dart, "draw did not increase vat.urns(ilk,vault).art by dart";
+    assert nstBalanceOfBufferAfter == nstBalanceOfBufferBefore + wad, "draw did not increase nst.balanceOf(buffer) by wad";
+    assert nstTotalSupplyAfter == nstTotalSupplyBefore + wad, "draw did not increase nst.totalSupply() by wad";
+}
+
+// Verify revert rules on draw
+rule draw_revert(uint256 wad) {
+    env e;
+
+    bool canCall = roles.canCall(ilk(), e.msg.sender, currentContract, to_bytes4(0x3b304147));
+    mathint wardsSender = wards(e.msg.sender);
+    mathint nstTotalSupply = nst.totalSupply();
+    mathint nstBalanceOfBuffer = nst.balanceOf(buffer());
+    require nstBalanceOfBuffer <= nstTotalSupply;
+    mathint vatInkVault; mathint vatArtVault;
+    vatInkVault, vatArtVault = vat.urns(ilk(), currentContract);
+    mathint duty = jug.duty();
+    require duty >= RAY();
+    mathint rho = jug.rho();
+    require to_mathint(e.block.timestamp) >= rho;
+    mathint rate = vat.rate() + (duty - RAY()) * (e.block.timestamp - jug.rho());
+    require rate > 0 && rate <= max_int256();
+    mathint dart = divUp(wad * RAY(), rate);
+    mathint vatDaiVault = vat.dai(currentContract);
+    mathint vatCanVaultNstJoin = vat.can(currentContract, nstJoin);
+    mathint vatDaiNstJoin = vat.dai(nstJoin);
+
+    draw@withrevert(e, wad);
+
+    bool revert1  = e.msg.value > 0;
+    bool revert2  = !canCall && wardsSender != 1;
+    bool revert3  = wad * RAY() > max_uint256;
+    bool revert4  = dart > max_int256();
+    bool revert5  = vatArtVault + dart > max_uint256;
+    bool revert6  = rate > max_int256();
+    bool revert7  = rate * dart > max_int256();
+    bool revert8  = vatDaiVault + rate * dart > max_uint256;
+    bool revert9  = vatCanVaultNstJoin != 1;
+    bool revert10 = vatDaiNstJoin + wad * RAY() > max_uint256;
+    bool revert11 = nstTotalSupply + wad > max_uint256;
+
+    assert revert1  => lastReverted, "revert1 failed";
+    assert revert2  => lastReverted, "revert2 failed";
+    assert revert3  => lastReverted, "revert3 failed";
+    assert revert4  => lastReverted, "revert4 failed";
+    assert revert5  => lastReverted, "revert5 failed";
+    assert revert6  => lastReverted, "revert6 failed";
+    assert revert7  => lastReverted, "revert7 failed";
+    assert revert8  => lastReverted, "revert8 failed";
+    assert revert9  => lastReverted, "revert9 failed";
+    assert revert10 => lastReverted, "revert10 failed";
+    assert revert11 => lastReverted, "revert11 failed";
+    assert lastReverted => revert1  || revert2 || revert3 ||
+                           revert4  || revert5 || revert6 ||
+                           revert7  || revert8 || revert9 ||
+                           revert10 || revert11, "Revert rules are not covering all the cases";
+}
```

Comment:
This spec seems to be missing rules for `wipe`?

Reply:
Oh, yeah it seems I left `wipe` to be done after timeouts on `draw` were solved and I forgot to do it. Working on this now.

Reply:
I've added the rules and pushed them, but `wipe_revert` is timing out in "Revert rules are not covering all the cases". Waiting for Certora's feedback.

Reply:
Spec is passing now using latest `certora-cli-beta`